parser.y

```diff
@@ -5083,6 +5083,22 @@ CastType:
 		x.Collate = charset.CollationBin
 		$$ = x
 	}
+|	"FLOAT" FloatOpt
+	{
+		x := types.NewFieldType(mysql.TypeFloat)
+		fopt := $2.(*ast.FloatOpt)
+		if fopt.Flen >= 54 {
+			yylex.AppendError(ErrTooBigPrecision.GenWithStackByArgs(fopt.Flen,"CAST",53))
+		}
+		if fopt.Flen >= 25 && fopt.Flen <= 53 && !parser.lexer.GetSQLMode().HasRealAsFloatMode() {
```

Comment:
I don't think `REAL_AS_FLOAT` as any effect on the result type of `CAST(… AS FLOAT)`. It is only relevant for `CAST(… AS REAL)` which isn't implemented in this PR. 

Also, the `Flen <= 53` condition is redundant.

```suggestion
		if fopt.Flen >= 25 {
```

----

Demo. Note that only the type of `CAST(… AS REAL)` is affected by `REAL_AS_FLOAT` mode.

<table><tr><th>SQL Mode</th><th>Resulting schema</th></tr><tr><td>

```sql
set sql_mode = '';
create table a as select 
  cast(5 as float(24)) f24, 
  cast(5 as float(53)) f53, 
  cast(5 as real) r, 
  cast(5 as float) f, 
  cast(5 as double) d;
show create table a;
```

</td><td>

```sql
CREATE TABLE `a` (
  `f24` float NOT NULL DEFAULT '0',
  `f53` double NOT NULL DEFAULT '0',
  `r` double NOT NULL DEFAULT '0',
  `f` float NOT NULL DEFAULT '0',
  `d` double NOT NULL DEFAULT '0'
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_general_ci
```

</td></tr><tr><td>

```sql
set sql_mode = 'real_as_float';
create table a as select 
  cast(5 as float(24)) f24, 
  cast(5 as float(53)) f53, 
  cast(5 as real) r, 
  cast(5 as float) f, 
  cast(5 as double) d;
show create table a;
```

</td><td>

```sql
CREATE TABLE `a` (
  `f24` float NOT NULL DEFAULT '0',
  `f53` double NOT NULL DEFAULT '0',
  `r` float NOT NULL DEFAULT '0',
  `f` float NOT NULL DEFAULT '0',
  `d` double NOT NULL DEFAULT '0'
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_general_ci 
```

</td></tr></table>

Reply:
You are right. And I update.
PTAL again. @kennytm